### PR TITLE
Set minimum version to 2.18 for all pubspec.yaml files

### PR DESCRIPTION
**Pull request by Stargator**
We last time we did a mass update of the version of Dart used was 2.12. With the release of 2.18, I think it's time for another update.

This does pose a risk of forcing students/users to upgrade to a newer version of Dart, but that's going to happen eventually.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -20,10 +20,10 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        tag: ['2.12']
+        tag: ['2.18']
 
     container:
-      image:  google/dart:${{ matrix.tag }}
+      image:  dart:${{ matrix.tag }}
```
Comment (devkabiir) [inline .github/workflows/test.yml:26]: Is this shorthand syntax for a dart age that github actions "knows" about or did google publish a top-level image for dart?

File: exercises/practice/difference-of-squares/.meta/lib/example.dart [hunk 2]
```diff
@@ -9,5 +9,5 @@ class DifferenceOfSquares {
 
   num _sum(int input) => input * (input + 1) ~/ 2;
 
-  List<int> _range(int length) => new List<int>.generate(length, (i) => i + 1);
+  List<num> _range(int length) => new List<num>.generate(length, (i) => i + 1);
```
Comment (devkabiir) [inline exercises/practice/difference-of-squares/.meta/lib/example.dart:12]: Why is that necessary?

**Review by devkabiir — APPROVED** [review]
LGTM :rocket:

Long time no see!, how have you been doing?

File: .github/workflows/test.yml (hunk 1 above)
Comment (Stargator) [reply .github/workflows/test.yml:26]: Google retired google/dart, it hasn't had a new version of dart in a while.

So yes dart Is the new top level.

File: exercises/practice/difference-of-squares/.meta/lib/example.dart (hunk 2 above)
Comment (Stargator) [reply exercises/practice/difference-of-squares/.meta/lib/example.dart:12]: The dart linter applied this change. Let me look up the lint rule for it.

Comment (Stargator) [reply exercises/practice/difference-of-squares/.meta/lib/example.dart:12]: Hmm, well I can't reproduce the output with `dart analyze`, `dart format`, `dart_style`, or `dart fix`.

Maybe it was my IDE? Anyhow, I'll revert it

**Review by devkabiir — APPROVED** [review]

**Comment by kytrinyx** [comment]
@Stargator Are we good to merge here? (I'm thinking getting this merged fairly quickly will be good, since it touches quite a few files, and any other changes would risk introducing conflicts).

**Comment by kytrinyx** [comment]
@devkabiir @Stargator what is the process that you like to follow for merging? Once something is approved, are we okay to merge?

**Comment by devkabiir** [comment]
> Once something is approved, are we okay to merge?

Generally Yes!, Sometimes I wait for the PR author to make any minor changes they want and let them merge it themselves. Since approving unblocks the merging. In this case, I was waiting for response from @Stargator, assuming there were some changes pending. It looks like though, no more changes are pending. This is good to merge.

**Comment by kytrinyx** [comment]
> We can always have subsequent PRs for more changes. Or revert this in case of any issues.

Agreed! Thank you, @devkabiir

**Pull request merged**